addons/pos_self_order/models/product_product.py

```diff
@@ -35,9 +39,16 @@ def _filter_applicable_attributes(self, attributes_by_ptal_id: Dict) -> List[Dic
     def _get_attributes(self, pos_config_sudo: PosConfig) -> List[Dict]:
         self.ensure_one()
 
-        attributes = self._filter_applicable_attributes(
-            self.env["pos.session"]._get_attributes_by_ptal_id()
-        )
+        attributes = self.env.context.get("cached_attributes_by_ptal_id")
+
+        if attributes is None:
+            attributes = self.env["pos.session"]._get_attributes_by_ptal_id()
+            attributes = self._filter_applicable_attributes(attributes)
+        else:
+            # Performance trick to avoid unnecessary calls to _get_attributes_by_ptal_id()
+            # Needs to be deep-copied because attributes is potentially mutated
+            attributes = deepcopy(self._filter_applicable_attributes(attributes))
+
```

Comment:
```suggestion
        attributes_by_ptal_id = self.env.context.get("cached_attributes_by_ptal_id")

        if attributes_by_ptal_id is None:
            attributes_by_ptal_id = self.env["pos.session"]._get_attributes_by_ptal_id()
            self = self.with_context(cached_attributes_by_ptal_id=attributes_by_ptal_id)

        # Needs to be deep-copied because attributes is potentially mutated
        attributes = deepcopy(self._filter_applicable_attributes(attributes_by_ptal_id))

```

Reply:
I see the need to deepcopy attributes regardless of if it was cached or not now, because the mutation may be present in future calls with the cached `attributes_by_ptal_id`. 

On line 46 of this suggestion, to add it to the cache if it wasn't present, I don't think this will work because it will only add it to the context of `self` for this function calls' scope, but that `self` will not stay for future calls, so the cache will still be `None`. (Testing confirmed this).

I thought this may be possible to do if you mutate `self`'s context directly, but the context is a frozendict, so you can't just add the key/value.

It could be possible if you re-assign `self.env.context` entirely like so:
```py
self.env.context = self.with_context(cached_attributes_by_ptal_id=attributes_by_ptal_id).env.context
```
And this seems to work well. But I don't know if this would be considered an anti-pattern and could have negative effects in other places that use context.

Reply:
I do see some other code that does something similar,
https://github.com/odoo/enterprise/blob/e7bc4a9b31ce471e25238ff22fa4fd151d831867/documents_project/models/project_project.py#L167
But they reset context back to what it used to be at the end. So not sure if leaving the new key in there would be fine or not.